collections/src/main/scala/strawman/collection/mutable/UnrolledBuffer.scala

```diff
@@ -0,0 +1,435 @@
+    this
+  }
+
+  def addOne(elem: T) = {
+    lastptr = lastptr.append(elem)
+    sz += 1
+    this
+  }
+
+  def clear(): Unit = {
+    headptr = newUnrolled
+    lastptr = headptr
+    sz = 0
+  }
+
+  def iterator: Iterator[T] = new AbstractIterator[T] {
+    var pos: Int = -1
+    var node: Unrolled[T] = headptr
+    scan()
+
+    private def scan(): Unit = {
+      pos += 1
+      while (pos >= node.size) {
+        pos = 0
+        node = node.next
+        if (node eq null) return
+      }
+    }
+    def hasNext = node ne null
+    def next = if (hasNext) {
+      val r = node.array(pos)
+      scan()
+      r
+    } else Iterator.empty.next()
+  }
+
+  // this should be faster than the iterator
+  override def foreach[U](f: T => U) = headptr.foreach(f)
+
+  def result = this
+
+  def length = sz
+
+  def apply(idx: Int) =
+    if (idx >= 0 && idx < sz) headptr(idx)
+    else throw new IndexOutOfBoundsException(idx.toString)
+
+  def update(idx: Int, newelem: T) =
+    if (idx >= 0 && idx < sz) headptr(idx) = newelem
+    else throw new IndexOutOfBoundsException(idx.toString)
+
+  def mapInPlace(f: T => T): this.type = {
+    headptr.mapInPlace(f)
+    this
+  }
+
+  def remove(idx: Int) =
+    if (idx >= 0 && idx < sz) {
+      sz -= 1
+      headptr.remove(idx, this)
+    } else throw new IndexOutOfBoundsException(idx.toString)
+
+  @tailrec final def remove(idx: Int, count: Int): Unit =
+    if (count > 0) {
+      remove(idx)
+      remove(idx, count-1)
+    }
+
+  def prepend(elem: T) = {
+    headptr = headptr prepend elem
+    sz += 1
+    this
+  }
+
+  def insert(idx: Int, elem: T): Unit =
+    insertAll(idx, elem :: Nil)
+
+  def insertAll(idx: Int, elems: IterableOnce[T]): Unit =
+    if (idx >= 0 && idx <= sz) {
+      sz += headptr.insertAll(idx, elems, this)
+    } else throw new IndexOutOfBoundsException(idx.toString)
+
+  def subtractOne(elem: T): this.type = {
+    headptr.subtractOne(elem, this)
+    this
+  }
+
+  def patchInPlace(from: Int, patch: strawman.collection.Seq[T], replaced: Int): this.type = {
+    remove(from, replaced)
+    insertAll(from, patch)
+    this
+  }
+
+  private def writeObject(out: java.io.ObjectOutputStream): Unit = {
+    out.defaultWriteObject
+    out writeInt sz
+    for (elem <- this) out writeObject elem
+  }
+
+  private def readObject(in: java.io.ObjectInputStream): Unit = {
+    in.defaultReadObject
+
+    val num = in.readInt
+
+    headPtr = newUnrolled
+    lastPtr = headPtr
+    sz = 0
+    var i = 0
+    while (i < num) {
+      this += in.readObject.asInstanceOf[T]
+      i += 1
+    }
+  }
+
+  override def clone(): UnrolledBuffer[T] = new UnrolledBuffer[T] ++= this
+
+  override def className = "UnrolledBuffer"
+}
+
+
+object UnrolledBuffer extends StrictOptimizedSeqFactory[UnrolledBuffer] {
+
+  def empty[A]: UnrolledBuffer[A] = empty(ClassTag.Any.asInstanceOf[ClassTag[A]])
+
+  def empty[A](implicit t: ClassTag[A]): UnrolledBuffer[A] = new UnrolledBuffer[A]
```

Comment:
What’s the advantage of having both overloads? (the one that uses `ClassTag.Any` and the one that uses the implicit `ClassTag[A]`)

Reply:
You want the methods with the `ClassTag` wherever possible but `IterableFactory` requires the ones without. This is not the first such collection type and the old collections have dedicated support for this. We should consider doing the same and adding a new factory type.

Reply:
if we write `empty[SomeType]` which overload is resolved? It depends on whether you have an implicit `ClassTag[SomeType]` in scope? Did you do some tests? (my intuition is that what’s is in the implicit scope doesn’t influence overloading resolution)

Reply:
Right, they're ambiguous in this case. We need to move the more general version into a superclass to disambiguate (like the overloads with `Ordering` in some collection types)

Reply:
Better ClassTag-based factories that could be used for `UnrolledBuffer`, too, in https://github.com/scala/collection-strawman/pull/391